From patchwork Tue Feb 23 19:19:55 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12100855
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 46E3CC433DB
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:11:57 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 0E70A64E7A
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:11:57 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S234191AbhBWULu (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 23 Feb 2021 15:11:50 -0500
Received: from userp2120.oracle.com ([156.151.31.85]:39738 "EHLO
        userp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S234177AbhBWUKE (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 23 Feb 2021 15:10:04 -0500
Received: from pps.filterd (userp2120.oracle.com [127.0.0.1])
        by userp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK4xJk023448;
        Tue, 23 Feb 2021 20:08:18 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=Zu8rfJyHRcTfVFgb/PRcCVioUNaAej5CPTn06OpwxaI=;
 b=ANYPDjznlJLRPt86883JwK154JQHPr4pQmUSP2/+mDP2AZbHRR9uswpMFlX3FUW4070U
 PJgFM44CkIVWK3RalIoPaeicnxAYgDRyI0UT3zRfCOkdSEKbX+cewLaxMOg/MBjEs3ZI
 FKrCR0PYGUMj/kWndtk9BdLot1Nx+n0hBTcBhk8re21bH1IjlUXbEFZPsZS2hk7tfU4H
 CuolpUtZom8UgBcHtD4nQSkgauyBueLsyE/uyyhurQO0T4C2NNIwUrw6AzpelYlXyldC
 6eBXJ4WQfpg31F1b5ObLGPrttcfnVQK1v1GiyVg4seihJNf2kT94eBwMB5c2TB758u5h 0A==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by userp2120.oracle.com with ESMTP id 36ugq3fjet-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:18 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK6Vd8106920;
        Tue, 23 Feb 2021 20:08:17 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2175.outbound.protection.outlook.com [104.47.55.175])
        by userp3030.oracle.com with ESMTP id 36ucbxyc3h-2
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:17 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=jgjtjiZo7w+ij7pFJ//t9ZHTsWDjaHtc/2TPrRz5evSIGfRO8bmfS1Hnt4pGxW4EWLJ65BF9DbhMFs80thDQWwR0KMwBl+a9LF1CT6q1TDUak1MY2RBpYbLKwJnDuVbg74BgBCDr/sKX614RUMOtHWYCaNQRjiLX6XGPGp562/c0mXS50XuvPoSto6dw+O/NicUHOYR2tTIaP3CPsdpao2faHQtNsAdmyOVfVPDJIyfjMcMGNKNRTWByiMehq4RuNQK10Hwx6hGg4npG0iA75MNIkw1XC+A9K52BJlp7QAubXRXRQ8NV3Km7vDid8KiqL5EW7bT7M5lXok0LGQVJ/w==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Zu8rfJyHRcTfVFgb/PRcCVioUNaAej5CPTn06OpwxaI=;
 b=ZFeZscCqPFdTWB8cfFJhIaDm2TlAyToJoPEr5B97nEhurY13mB8GwpWjdYgHoV1VefwYSBcs/eOOSLpUx0bfmGd3K3FJOvOazB0qpIRWe4uv78rZJScsSFFqWZoAHrWw4PnzB3Xv4FLhA2vuX+9Ef2fLYBnxVzqouo4uxYlNh4OA/7w/mYT3y5iO0/tGFJH4HdBRoFb3WOx4f3UdFiT6HoeEiLqWTJZDNz5vDuoYaa90Pk6TwNOe9SqzmLEVZtN85WiYKny3m4qJSINL79Gw2fTbJfQAstsMGN2JsoVoV/c2VnHCAPk4BwaUOIwnY9VmFD81PGSvv77vsybt6bNyQA==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Zu8rfJyHRcTfVFgb/PRcCVioUNaAej5CPTn06OpwxaI=;
 b=ziyYwZpEU+OkqrWTt+fNkCxy3S3k9JFWeKhci1bZPIxLidjh5d91jfCtRI92fAjBmpqJDt5mJBd8JvZ8q80nqbgq7hqoB6nrOyXvJzbhI3P1BLNnMuwf0hvqD0hamq0c/E98pAVL9RZB1NhRBJ8NehEz4psOqpLA5YcwG1Oy59c=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3868.27; Tue, 23 Feb
 2021 20:08:15 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3868.033; Tue, 23 Feb 2021
 20:08:15 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 1/4 v3] KVM: nSVM: Do not advance RIP following VMRUN
 completion if the latter is single-stepped
Date: Tue, 23 Feb 2021 14:19:55 -0500
Message-Id: <20210223191958.24218-2-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
References: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: BY3PR05CA0026.namprd05.prod.outlook.com
 (2603:10b6:a03:254::31) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 BY3PR05CA0026.namprd05.prod.outlook.com (2603:10b6:a03:254::31) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3890.10 via Frontend
 Transport; Tue, 23 Feb 2021 20:08:14 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: db3ffe92-c370-4361-616c-08d8d836c106
X-MS-TrafficTypeDiagnostic: SN6PR10MB3022:
X-Microsoft-Antispam-PRVS: 
 <SN6PR10MB30222E30669C7853B388C24F81809@SN6PR10MB3022.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:7691;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 A64eL3FuqI/pllC4pzUmuPnmuv5wW39/DtDeeWv04byg6UxALVmPqCf6noSuslpzgC7G10x38vjq7dGuAOHDnkkwUmBd2DpaeqjhqdBOB40rEgQ+TqzqMa53zIvGnn/sLCJ9Acf+Pc0GfGEwK6liFznxjAWF5UDA03T+Sw1AcEecxMUWlbJXkTbojcjE/S4BDa0lyha971JlPQ/fCPUCP8QBDRSzr3gUuVrx0EZc6N4KsrHyRtotZRtG2OzaE/H1yJvJL6wPzOj0WNWBhwAn10doNx0fQFBjZsCevNHFhujAqDCH4wSuMesgewy9hhFViX3y4SNHHkAnfU7o1mFMgtUYyH7NiPcyQ6HYeZ+/lYByhiFQ2Q5uu87T5snre5BGH0AvyGjnGl6IMTvSeEcCGra8cie/r6VqKeDQGoTVgtWgi/eSmthDMYtybXgfxlKL1uX9MdQadi2Bbx4LpIO9futYU0rL4dlULQ6OJqF8j9jztV1PdMw/zU1dCcI0REv+In9Gd9uFFj6ZQJDeiRwvEw==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(39860400002)(346002)(366004)(396003)(376002)(136003)(66946007)(26005)(86362001)(7696005)(66476007)(66556008)(52116002)(1076003)(186003)(44832011)(16526019)(36756003)(6666004)(316002)(956004)(2616005)(6916009)(5660300002)(6486002)(4326008)(83380400001)(2906002)(478600001)(8676002)(8936002);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 gV8XunApVChU+Vezzb5bPLR/zsu+Ekb19/BVSqLAU3goX1mtPAhi0JMK9rdgcg2WjXdXa3K5rnriqGNib+HkeJTUL+XW8FOlXrEhls1y65370Y9US0JjxVkcw7VcbZhGW4H+r/gdZOLXz+nkEQOpQy5CXsk24sQgGJCcU2yeOzlvOlx22HQOsLkpmER1EVjJx1ehCzWX8gILrsBH8c/O2wy+jYRvY+GXgx0AQwPkP+0lGEL5fdkYgxxGm9PDC46xMUke+7nDQ441wdGzjykMpN9k1jCDppMd04WTFC1Ojniuj15zmQWleJQLj/OPdxa31AjK8hNIVfBdJOXAGPn27520zxQud0detEt6pN1oY/4r3kCOhj/RHf+TUItMBdFk0PWO1sTsayBspD6bUAOtoQfW4X8VtAmZbbv3ID0RBToqolW6fawIQhmE6HNxql4TpyjuDN9amZw5ODOiq70A91iSFAIdKKiCzdWWV+j+441ozbFk6hbedEMK1NiNKkN6Oih/8+6NECfMAzsq9xEUN97wgQF8R71+TlaljmFYAPgR5OTQ2fTurauePdWKTmKd543N1iYe4gWfxoeL1CW/wVTCeBA/R2sY/J3SsAzq9EkW0RiJj8CVdg7sJToO1A+yHrIY6zJZGihIXM0/mWcH+vOdpHxMEt/0sJZSo47EV6AdJ3S+f+FEB6nZjC+O2tcud4nYRBDP/kgv/CX99ETosXg3IiX+G1bNwYERNiYJvrLJga2wT7zaoUP1Ac4DEpuAmkGzRSFRCxxO2MpkyF1mrPGO/XVIQNSp0xZVg2CTo+voBgvwxp+8e0xZhEfaOqHs+8DZgTh9KCkNzSxaPFhv2UDLAqyNw/Jq7gwOhXU17a+vkGCCtzcEsjvzK24LmiGFbUtdM6e05boxWOmJaxpXW7VBiUOewhnkC/BZwWz59ZpmTuzyJVND0IonmjTCCLsu/p017PVBhW+3Ew4DFAHjbsAiCYU89pfdiaiLmT/uFJ+wWAej8RZJoyAddz8jBocnQO9PeR4y88YEoAkk46Woq+2QDgySKyXlyItZWEk1kp5cXyAplccxS7m5duC4iObjco/YGtIhFRz4G/UnEM1cRCHa+4XD7cW8VsiPRYVuqHKJtHNllUToUH65dmi9WJQtctFTI/m0kLjj9GwOdmZBALAUmTb6cn29iPl2l5pZndrLM0sUEwyNf74jZ/IS4GjM+reuLCO83S3+uPH4wbMdHn2JVqj7m9q0w6sgPI6wn0pXsuBcOrNS/yF/UqNry/7ozu5l3D5t89d46e28GATE2VghgeUbNZn5/wj7B2gPEEW8Vz/d+dkg+Vc7GZZKuwKC
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 db3ffe92-c370-4361-616c-08d8d836c106
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Feb 2021 20:08:15.3528
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 2ysVrlQm7yeTPbQmTQGWdGLVL2+mPp2q7hT2YmD800NG+BbRYy8eOy/YPc3J1wXvNa3Qwo2t+rlGQfBQjpHadC/3YvT/xhKNs0KglEZU66s=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR10MB3022
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxlogscore=846
 adultscore=0
 phishscore=0 spamscore=0 suspectscore=0 bulkscore=0 malwarescore=0
 mlxscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 adultscore=0
 phishscore=0
 malwarescore=0 spamscore=0 mlxscore=0 suspectscore=0 priorityscore=1501
 clxscore=1015 impostorscore=0 lowpriorityscore=0 mlxlogscore=999
 bulkscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Currently, svm_vcpu_run() advances the RIP following VMRUN completion when
control returns to host. This works fine if there is no trap flag set
on the VMRUN instruction i.e., if VMRUN is not single-stepped. But if
VMRUN is single-stepped, this advancement of the RIP leads to an incorrect
RIP in the #DB handler invoked for the single-step trap. Therefore, check
if the VMRUN instruction is single-stepped and if so, do not advance the RIP
when the #DB intercept #VMEXIT happens.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oraacle.com>
---
 arch/x86/kvm/svm/svm.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/arch/x86/kvm/svm/svm.c b/arch/x86/kvm/svm/svm.c
index 3442d44ca53b..427d32213f51 100644
--- a/arch/x86/kvm/svm/svm.c
+++ b/arch/x86/kvm/svm/svm.c
@@ -3740,6 +3740,8 @@ static noinstr void svm_vcpu_enter_exit(struct kvm_vcpu *vcpu,
 	instrumentation_end();
 }
 
+static bool single_step_vmrun = false;
+
 static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 {
 	struct vcpu_svm *svm = to_svm(vcpu);
@@ -3800,6 +3802,10 @@ static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 
 	svm_vcpu_enter_exit(vcpu, svm);
 
+	if (svm->vmcb->control.exit_code == SVM_EXIT_VMRUN &&
+	    (svm->vmcb->save.rflags & X86_EFLAGS_TF))
+                single_step_vmrun = true;
+
 	/*
 	 * We do not use IBRS in the kernel. If this vCPU has used the
 	 * SPEC_CTRL MSR it may have left it on; save the value and
@@ -3827,7 +3833,11 @@ static __no_kcsan fastpath_t svm_vcpu_run(struct kvm_vcpu *vcpu)
 		vcpu->arch.cr2 = svm->vmcb->save.cr2;
 		vcpu->arch.regs[VCPU_REGS_RAX] = svm->vmcb->save.rax;
 		vcpu->arch.regs[VCPU_REGS_RSP] = svm->vmcb->save.rsp;
-		vcpu->arch.regs[VCPU_REGS_RIP] = svm->vmcb->save.rip;
+		if (single_step_vmrun && svm->vmcb->control.exit_code ==
+		    SVM_EXIT_EXCP_BASE + DB_VECTOR)
+			single_step_vmrun = false;
+		else
+			vcpu->arch.regs[VCPU_REGS_RIP] = svm->vmcb->save.rip;
 	}
 
 	if (unlikely(svm->vmcb->control.exit_code == SVM_EXIT_NMI))

From patchwork Tue Feb 23 19:19:56 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12100825
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 4C932C433E0
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:10:49 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 06EAA64E6C
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:10:49 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S231818AbhBWUKj (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 23 Feb 2021 15:10:39 -0500
Received: from userp2130.oracle.com ([156.151.31.86]:51706 "EHLO
        userp2130.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S234170AbhBWUJK (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 23 Feb 2021 15:09:10 -0500
Received: from pps.filterd (userp2130.oracle.com [127.0.0.1])
        by userp2130.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK5aov172602;
        Tue, 23 Feb 2021 20:08:19 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=K+cmskWO4KMrstqQI6PxoqOUASU42S4XbV+9Yen6tHk=;
 b=IpSwoGQ4PJVvyjQ9sQqhMqlYfEpQKqe2PUh4dF+eTT9xnyVJl7KVzjFQdKv+YtwfUDSl
 g06EKeSBo1BRLxrz5mfKbI5n/bhLCqovsnsStwK2pZSfd1Aytff9pqyOGUyE6FGPei2f
 gWr+/78iuNjN52VaCoHRAFE/+0yU5fnJOaYJ3sATOGb8db3TrWXWNZwaMlOe2sUKqqcc
 jTtBlv1M4ANM0PPns7xz1cGveEPX7aRUqKH+05omxXYnrBySL9ayW5mz/TYHwvE5SI6x
 Z34rHow+h2WPvXullOOO5+m+0JklmdWkfqWyYo9FIeusxwFXFRTiZqmLzekwvNzT3C8Z qg==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by userp2130.oracle.com with ESMTP id 36tsur0qtm-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:18 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK6Vd9106920;
        Tue, 23 Feb 2021 20:08:18 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2175.outbound.protection.outlook.com [104.47.55.175])
        by userp3030.oracle.com with ESMTP id 36ucbxyc3h-3
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:18 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=PzDmudYvxzeKaS5p4XcGC3fqojZjsnu0p+3tHrnl9hYoYq4KfzHc1TDNwQCqOGlJ3dcITPICGMHcwdxeWguiFd3OipAP/3yxXj10kyDPhIYWI3unHaAUUt77cW/tZeaqc4F7zmc55E+eXCyUwM5fQkH7vWnQim1e+qySx5eE1wPy5RYEWmHzy/Mfpv3P2GPqkrIZjhg75rLqbC/pIuZjfqPWEiZh5BbKhLXpWQbL9B3VN36mS/JfrjZcYXLHu2+2SrjI/NuHS+0lH3aTCjf7jQO8flOFHUvlCmfhkIwR8RB4k6//WSm3hnIabjEidElpvW2QloWUt7u/U18whYc+uw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=K+cmskWO4KMrstqQI6PxoqOUASU42S4XbV+9Yen6tHk=;
 b=SdDPDpg3wc2bIMM1QM46qIZpyPHABxbycNwEk3+xkDjOntQ8tqFnedI7NGNXGWescyT2Om+nD7CwoSIrb6qMM+PnuiIENyQcGy2btAQQjJEWrwrCHNvtY75amAGqMXH5cDZq4lRs+WnR3KRAV2+JfaNtzPxyOk+L8Lpo1Lkjg+xRvwNs+mZ6pb3hX6jrECnfodnP1XvI5oy3Khl7oDNp7WHQUN2mOzzRESRniqfsJ54m3eJURsYKJ8h+rz4d76gKcxXt3floCo5ld8ZYwUfxcX8n3+7Z+4UcJf0J22Ido/egIzjDatvOWIYYwQP87IbI6/P8a9kgdBfrAt9Bxw36gg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=K+cmskWO4KMrstqQI6PxoqOUASU42S4XbV+9Yen6tHk=;
 b=V64+O7vSBkdQcrNUL8z82j3BY53RRMdtf6SkwNd7wzIyl9tyiu853qBiUbchhirdXRr9B8WRNKwjfKHRmKOmrlOql1VK9qHLYiFPoxbq9k1b3DgxtQrBIvi1XXe6jVYXOd7Whl+Qc6irQBw0JGsXrUlFmQno8RzpqbQqSqUg+VA=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3868.27; Tue, 23 Feb
 2021 20:08:16 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3868.033; Tue, 23 Feb 2021
 20:08:16 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 2/4 v3] KVM: X86: Add a utility function to read current RIP
Date: Tue, 23 Feb 2021 14:19:56 -0500
Message-Id: <20210223191958.24218-3-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
References: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: BY3PR05CA0026.namprd05.prod.outlook.com
 (2603:10b6:a03:254::31) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 BY3PR05CA0026.namprd05.prod.outlook.com (2603:10b6:a03:254::31) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3890.10 via Frontend
 Transport; Tue, 23 Feb 2021 20:08:15 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 462dcdef-d778-44e0-80ae-08d8d836c1af
X-MS-TrafficTypeDiagnostic: SN6PR10MB3022:
X-Microsoft-Antispam-PRVS: 
 <SN6PR10MB3022FF2CFE76AC579D75E24281809@SN6PR10MB3022.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:6108;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 qjfglLfse/w+RNcIHOdsAP/nbOkA/WWElUPf7eTR+g+2i6wimujZ7+4xwOsAyH78svcYFvSeph27JhSCbfooy68Jk4bT1gLgS2qQKqMMzoX2HWwv7ztED3fYlhzfh9ENxv6rGD8Tkpoay1Ynq+JQ6778DrWtJdOqkpwIMxmAvFiqamp52dFMdZ0HgIcWQWi9ps4FpilgJGdhlLlvn5sEG7QkbdADRJJIofPK8bsLF5u5I8BKcqT6w70lXRHI/zVH6NiyV0tCkxKGzcOtvpCXc964c+0w9LqUU3XercX3galcTW+LCvgcuJGLJv+SwSWepL3JIzByUBnCYRFl2n7hRPwgpHJAPf7zPAPKqXzyaStk9ytgIYpxn3Vn7rGjNFnrplMt4HVIU0/tvsQ7HgigqTsvmY+IX52S4JPJVZ1iMSL7Zj8/Ru6DT7NXK9vdBb+L9TSbcuGvO4hNLfbGABMkncaUaihIPMiEmoPoZo1EoHbGdCYiq9hPv50U0BlP/bPM0SaZvN8UkLwc9G552RqvpA==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(39860400002)(346002)(366004)(396003)(376002)(136003)(66946007)(26005)(86362001)(7696005)(66476007)(66556008)(52116002)(1076003)(186003)(44832011)(16526019)(36756003)(6666004)(316002)(4744005)(956004)(2616005)(6916009)(5660300002)(6486002)(4326008)(83380400001)(2906002)(478600001)(8676002)(8936002);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 chjvBEzeJG5JiIL+D/z9/bIq/feP+skpVdXQqwI/A8mpHB83sIDscZ3oM+qpI5UZMJ+sDJmSRJH4TWCVc7gETl8Y/RdGcn+GAIn+0j7yJqO34AUtk6+YAo7wFTp/khhHMzdIAHzJROF57afUcVMtNd/saQxHr6BUB2MmK8zjyAK3uH71LS33mVkwPIwghz+9hWJtj23kShoPEtSQnRyiH+Rkomm9Q4w1+EfMvOgOqUYN5Qbi1J5AnX/JPZvntHMgZWboxuuthIAvP6SOg/vnr2R9irF7NI0KnfhJjWkEM8oebNYtFU8pAijeLIOrH41+uMR1k9gyQgnNq0Kfg81s0foXRCsGLINHvkgwx2wDfPpyAD/1MIQly2BC5sxrpUNkw9W1PPluzNhnl9xAXAOibw/TaytEpdMOE05FR4048XlGlNMZ7r2jqwHiC/IKTdGIFPtOQ/X6vFW/wirlOACVLPmhAHPQQlfcgmopP2kFQoq2rWDxfVFmKI4Ybc9AXbTrMIgsgxGeqroKDn3hGwfv0eo3kAAwA2DdhqIwu0twiRU6cDYlwAyd+hTVyaANttUe90nad2ByHdF7Bx8CJX06wZhOkTUSTskcC6GM/WbCLOzmRP5uE8MD3cnz1IJzXJ8FHDsUZA62bE76Yk1edtahgfMd+eWTwSraPt4hM4IwWHNDwJDoowgqZCGlSNpzo6/DVAWxrZSdQg3GvqUtvUHlkILv7JhYTcDy2tA1dzFLMR4w6CbG78YdHeKZWas3mg+vEfCnYIvLHQLUOq7v6zRVK14uCdJdOU+UMKNgtp+jItArXS4B0wh6jH9OX6UJf834kyWldHxBJ8XqyrC4/O5lxPSjG2WZYGCDdKZJGFwpZ51cOESMpL7dLmhbIkjRlO8QzeJr+e9289NTv9PygzNMYlbXQoglgPFDa9HAJPWnbE+EWsMljHoOFkEI00LelUV9oCbP1qrUG6S+nHduIBoSlyVmsJRsgpP3k4OnD6QuxKVtUIH6Pd9AvFhN9+aB4da/Erh36aMkmTVqJYnodbPwBmZQaVyIj6CZnrFrquX49s7OmMUfkTY1C6PDFu6g/ZgexjB90YDK92VmCGxevEdaLKsk8DNlpsWpgLPnrSGBPM1ihb2cgcqhL4HqRpKG5ayeW1ZCYcPAmVfnqlKKYiyhzo24JQC3N8zzaDLYCMgVZzCN/kIB2fv9UPzLaphN/ell+uKQeqLxu9kfErY+2zbG0dHygAsjfLr3zPlE9Fhn1y4k4hOXfR91uRFlqaqv9n36PBMx18/XHGQhSh+1NiaUJEU51HOSuH3A578S3KNariyYowUrr3fK3gsu47DtcAv7
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 462dcdef-d778-44e0-80ae-08d8d836c1af
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Feb 2021 20:08:16.5071
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 IcF3anRUl/ssrPEUpmHYUqfgtrsTgfbNd5tlhEPDUz1e7kYRqPA1XP8zkjyP49pnZp28q5kSanpxOmjZ4C5snGDASVVUOzEWFxjoA2eUdKE=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR10MB3022
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxlogscore=999
 adultscore=0
 phishscore=0 spamscore=0 suspectscore=0 bulkscore=0 malwarescore=0
 mlxscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 suspectscore=0
 spamscore=0
 priorityscore=1501 impostorscore=0 bulkscore=0 mlxscore=0 malwarescore=0
 clxscore=1015 phishscore=0 mlxlogscore=999 lowpriorityscore=0 adultscore=0
 classifier=spam adjust=0 reason=mlx scancount=1 engine=8.12.0-2009150000
 definitions=main-2102230169
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Some tests may require to know the current RIP to compares tests results. Add
a function to read current RIP. This will be used by a test in this series.

Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 lib/x86/processor.h | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/lib/x86/processor.h b/lib/x86/processor.h
index 87112bd..5604874 100644
--- a/lib/x86/processor.h
+++ b/lib/x86/processor.h
@@ -489,6 +489,13 @@ static inline unsigned long long rdtsc(void)
 	return r;
 }
 
+static inline u64 read_rip(void)
+{
+	u64 val;
+	asm volatile ("lea (%%rip),%0" : "=r"(val) : : "memory");
+	return val;
+}
+
 /*
  * Per the advice in the SDM, volume 2, the sequence "mfence; lfence"
  * executed immediately before rdtsc ensures that rdtsc will be

From patchwork Tue Feb 23 19:19:57 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12100823
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,USER_AGENT_GIT autolearn=ham autolearn_force=no version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 215CAC433E6
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:10:37 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id EDD6F64E7A
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:10:36 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S234201AbhBWUKZ (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 23 Feb 2021 15:10:25 -0500
Received: from aserp2120.oracle.com ([141.146.126.78]:41240 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S234169AbhBWUJE (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 23 Feb 2021 15:09:04 -0500
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK3woE091375;
        Tue, 23 Feb 2021 20:08:20 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=X6z1gAXO8hMUp5ADhTwYC5JrmdLMsJ/F4loi4h97qzc=;
 b=djGLKissKpyKut/RZzlTe/MGWemZC9UjhpYxaowAJUFUGRJ96gpuPt7ix38vWuzXUX/Z
 gtT0NSV2+QqxxVGymyPKQnatv4MO6SNH+xmCN7ER3VKB2EHiAMa9+YmRuIl6aLklpUru
 l/P4jSj49+O1dD7pgdN8sfMt9nMBk4P8Ar/kEoAFtQ+9UA6tRBhx9NMz4uW2S4qjjV+e
 F3zgdLWTp/gsJkgOYXCD1HAfGXD7w+nO09+8vlHjcBIlkpOiUhhmmF0dy7wondEtJlL+
 cZXNEG+sQY5FH+IyrxGJ/KkE4L77MT5+9nq2uV349Rdzs3y/lnS3LVeITgrYDVZOKCV4 uw==
Received: from userp3030.oracle.com (userp3030.oracle.com [156.151.31.80])
        by aserp2120.oracle.com with ESMTP id 36ttcm8pm4-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:20 +0000
Received: from pps.filterd (userp3030.oracle.com [127.0.0.1])
        by userp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK6VdA106920;
        Tue, 23 Feb 2021 20:08:19 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2175.outbound.protection.outlook.com [104.47.55.175])
        by userp3030.oracle.com with ESMTP id 36ucbxyc3h-4
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:19 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=aWtQJACA/fwgf1ZfuXzls77J4LNgaJ2/gPzC4D1yzl4QJhLbsGYQKs4Hd6uXRaq4+5bMvMaBrKKiWqmTWWA/svMk0ZcgFsE8lU/zu/FOyT0cWuAg7Qh6U9qmum1N/AmxHVHyov0wnUpvZxBPfJ551yvCK0NdBIBPZ4CDQvsCRO545TDnuhfluo1D80dFdOkQnKV4w+j2CiC1K6bPv/iB8izoFY5AK1L2EPO55Tss+cTRkpDWT3SFbvwdJxp736frtsfkvjV/zDgztyM5bjJ5M9yHgqFrOSbKvhnj83++njU4EshvHHVhS794kqr5Ur5H7ekaSi7qx7kgOL2qQLbUiQ==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=X6z1gAXO8hMUp5ADhTwYC5JrmdLMsJ/F4loi4h97qzc=;
 b=IeA+eicm/lM3TiyWwg6BhW155jAI9Lyb5NQlv9279VMP3Dp2Ts1oxoSIjlhaWK9IVcxQRg54qIiWyYRsdYJQNSkzK8iJEwcSfWountcdcCl1wy+jydX5udw+5w00g0VeiSsiEFF+m1a739c0nJ0Ellggo62Jd27FAJyZM8mK9dyzW0cxMGvqIUk5SezxiaiUnftw3skNvvZaMfUFleYQD90A41VEPyxP+20t8kG3n/6j2JPbkKA9PtKmYvOd9O8G3dYUzx4iA74MfaWH8A2XpeQEdvRyk/vT3eXb/UZLwGzfGTpBE8IvVk/jLhk1KEHM9p8XHP7lL4XdYi1YNuCKrg==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=X6z1gAXO8hMUp5ADhTwYC5JrmdLMsJ/F4loi4h97qzc=;
 b=GcVQOhl64DyYyKDLn1R71n2sQPGmjSsYVZ+LS1zXdbEfG+NrSF4rzopAdTohm+FHr4d6DgOONfZDMN2i1ck7iY0T83gDXHHZyrICROa4N+2cdRVWOtws7EeLLznLQTl1zemh/iY18jMVeVfMhRLKzN052FSZIts2yfHW9wNsqAQ=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SN6PR10MB3022.namprd10.prod.outlook.com (2603:10b6:805:d8::25) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3868.27; Tue, 23 Feb
 2021 20:08:17 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3868.033; Tue, 23 Feb 2021
 20:08:17 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 3/4 v3] KVM: nSVM: Add assembly label to VMRUN instruction
Date: Tue, 23 Feb 2021 14:19:57 -0500
Message-Id: <20210223191958.24218-4-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
References: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: BY3PR05CA0026.namprd05.prod.outlook.com
 (2603:10b6:a03:254::31) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 BY3PR05CA0026.namprd05.prod.outlook.com (2603:10b6:a03:254::31) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3890.10 via Frontend
 Transport; Tue, 23 Feb 2021 20:08:16 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: d639a3bd-d187-4333-9ea7-08d8d836c26a
X-MS-TrafficTypeDiagnostic: SN6PR10MB3022:
X-Microsoft-Antispam-PRVS: 
 <SN6PR10MB3022D0EC982871B7FA0D3A9D81809@SN6PR10MB3022.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:4125;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 n1fnfW1htOIZc32FkXITFrHbLQ92ZgcyGkldwmS03Akn+sCCP04AS7VTs5zv9iRJ/PWhWZqqLKNzpKArbNDeE8VEAoJVkO8ph+qcIl7OUKr/7+sAmmHf5V8f678qxInx4qZoG3DdaX8hRFSsXw3kqV9gBLkQePea2rb7Ae5fNZAuV2vv43aghLrxOR8DUEryrsr6CUhwWqdFzTznYXw/8VCtsKtgVYJ0WUyNtGAOHKHoIhw18BDVvdNW3ai1KDQKaA1byoN2g4akRiw7FqhN+xcHV1T1I4ZaXLMemvYK/fV6twnHM9vWkrXiOrkdUysv3hPdFx9I+N/Jf/1+FFQeugYr7SambY4OUTyv6TixTM45zqtzvb2n+tWkslqDENx2IH/RX1hWAAM1aRYGAeFzDm5fWCg/0qxi+EGnPxVOlqjd9BIaYNO6mfTCK7oLhaAlQkrcIzwIZzWfZeJ8iEzD7i39YJy01qVF98cV0nAyi84Lor38ri+UgZ3cnVS/uqb+BfLIU5I0pgmHYPfR0gOdUg==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(39860400002)(346002)(366004)(396003)(376002)(136003)(66946007)(26005)(86362001)(7696005)(66476007)(66556008)(52116002)(1076003)(186003)(44832011)(16526019)(36756003)(6666004)(316002)(956004)(2616005)(6916009)(5660300002)(6486002)(4326008)(83380400001)(2906002)(478600001)(8676002)(8936002);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 eQE/FEZRPQVNUIi57ycInasCKuxghOsZWqgrYK5LrSI5CFMuOOSbu8913aTeLKuMa3KCLVn77cuM3ikcsiyzbkHT4wLNU+t+ko2p9aAdtvQU2irdSJo3zs2algJBBSyjtThcgh3O7NndRYyJDFtsd0NwQWbBmbeafW0XcdplETiW2Kf6igS6a2T4ashrnuX6N5yVmOqy5DBd0qF1gkRagWti7vpVhFOmjEjs3P4M2i3uPnh4DKue+y7HABqihh6WBNaJ6x2RJD+oQ2lo6+vOJGn6LAYMkfWlge3/atbspwtv5Fuz5IFDB9/cqfelP1f/l9rBMbQwXeKIjCRz2EuIXMSgLIUHns/vt9i/WuaZyGnj0o5UTakJVi2MAHTVw1DH0Q3O0Yw1X7khzo8leGL8WROqtlZoXZZ4N/DFiD1qFW+PI7Ot57ZFCK7szdrWPWzRyYU4XfrXROq9mnaxZwIj9LJwL4cdbApYbydMB0DTo5bjSe1/4XDTaZkedgHKXRSAQxwJxbEJLjnQKZQG7UPjwrU7GN7M9peuf3mRhE+uvyAB9Jc5GAP4pyyii7WD/sxKohOHqijwZ9zY8Pm/dVdUfnQhIjKOGIAc3UMcF2VRti9oRgi7TSN2LiisWxXKPpXjJYX6yWx/RYdDxoAMARovD/RzTYMdwcwoekDJl0ANnK+SUhcOKnY8kjqGBb6EJ/+cRQFIDcsmvepoe5HdqnQ7korNw16biEw5PwOdTxU739E4iGJFa8wnZnOrhP09HblVacvbn8BwTaRfPbJk6wDtjOB2+wRxeIsGuJkPFIR0Y42qsP/JoDl7ieyqfgPuzU4AgXSAOIMa0z3bY6zPxttuN1e2xd6SbfxghYfnzxh8OQhfnFI0ewNQ/9oP0fXlm1gGkqDF4ytMwNBjTTmy/AcxTcRpGy2ypWeBxWHSyEj3MVaLp3byT6t0eYctAWH6yVPZxtB7Y5Dq9t2ujmC4Z+YsAaRvI7lKc6dZZnTf73XjA8t15lRAY9a5K2lCX7gfEYkh0u+aZJUqu7MixOt5IR4kJWrESeHnGD73IfH3ghScENDg2aUafmJ2J/cgmDQo80uJrFDSifmxRmHoJ84yABgePVl8oFrKBCOgbj6hyJGAsVrOshIONyVwjBzQoQ7jU8v2s9kmbe6BrdzVt5eGihB184o6HGHOsJEP6vnHJq0zAKGA1j//hrsXBw9kQ/HVc3DCtMc6Fzt6IC43aOKbXR0gWuH/3eu00TPrw3jHlb5Gq906u5pS9RGVBIS/HOkWJEIRWvx+KXW8vvjX9XV0eBCMg0txLO/Z2ok5OathIPPBS3Ux9f/0J5N51FldaJuthnzh
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 d639a3bd-d187-4333-9ea7-08d8d836c26a
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Feb 2021 20:08:17.6964
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 ubuDQ1u0vIVLFrXDlbwRWGXq7rw/mxDDxYfFnBMs0Hc6KDkY2e2THfWQicU4BvEDJiyZ8HA6v41uBZfwJpmo+MPVKO2F0RXSljoRxmIUKLI=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SN6PR10MB3022
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxlogscore=999
 adultscore=0
 phishscore=0 spamscore=0 suspectscore=0 bulkscore=0 malwarescore=0
 mlxscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxlogscore=999
 adultscore=0
 lowpriorityscore=0 spamscore=0 mlxscore=0 bulkscore=0 clxscore=1015
 priorityscore=1501 malwarescore=0 impostorscore=0 suspectscore=0
 phishscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Add an assembly label to the VMRUN instruction so that its RIP can be known
to test cases. This will be used by the test in the next patch.

Suggested-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/svm.c | 16 ++++++++++++----
 1 file changed, 12 insertions(+), 4 deletions(-)

diff --git a/x86/svm.c b/x86/svm.c
index a1808c7..77fba8b 100644
--- a/x86/svm.c
+++ b/x86/svm.c
@@ -208,14 +208,15 @@ struct regs get_regs(void)
 
 struct svm_test *v2_test;
 
-#define ASM_VMRUN_CMD                           \
+#define ASM_PRE_VMRUN_CMD                       \
                 "vmload %%rax\n\t"              \
                 "mov regs+0x80, %%r15\n\t"      \
                 "mov %%r15, 0x170(%%rax)\n\t"   \
                 "mov regs, %%r15\n\t"           \
                 "mov %%r15, 0x1f8(%%rax)\n\t"   \
                 LOAD_GPR_C                      \
-                "vmrun %%rax\n\t"               \
+
+#define ASM_POST_VMRUN_CMD                      \
                 SAVE_GPR_C                      \
                 "mov 0x170(%%rax), %%r15\n\t"   \
                 "mov %%r15, regs+0x80\n\t"      \
@@ -232,7 +233,9 @@ int svm_vmrun(void)
 	regs.rdi = (ulong)v2_test;
 
 	asm volatile (
-		ASM_VMRUN_CMD
+		ASM_PRE_VMRUN_CMD
+                "vmrun %%rax\n\t"               \
+		ASM_POST_VMRUN_CMD
 		:
 		: "a" (virt_to_phys(vmcb))
 		: "memory", "r15");
@@ -240,6 +243,8 @@ int svm_vmrun(void)
 	return (vmcb->control.exit_code);
 }
 
+extern void *vmrun_rip;
+
 static void test_run(struct svm_test *test)
 {
 	u64 vmcb_phys = virt_to_phys(vmcb);
@@ -258,7 +263,10 @@ static void test_run(struct svm_test *test)
 			"sti \n\t"
 			"call *%c[PREPARE_GIF_CLEAR](%[test]) \n \t"
 			"mov %[vmcb_phys], %%rax \n\t"
-			ASM_VMRUN_CMD
+			ASM_PRE_VMRUN_CMD
+			".global vmrun_rip\n\t"		\
+			"vmrun_rip: vmrun %%rax\n\t"    \
+			ASM_POST_VMRUN_CMD
 			"cli \n\t"
 			"stgi"
 			: // inputs clobbered by the guest:

From patchwork Tue Feb 23 19:19:58 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Krish Sadhukhan <krish.sadhukhan@oracle.com>
X-Patchwork-Id: 12100827
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-18.7 required=3.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,INCLUDES_CR_TRAILER,
	INCLUDES_PATCH,MAILING_LIST_MULTI,MSGID_FROM_MTA_HEADER,SPF_HELO_NONE,
	SPF_PASS,URIBL_BLOCKED,USER_AGENT_GIT autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 5007AC433E0
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:11:23 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 1ADAD64E6B
	for <kvm@archiver.kernel.org>; Tue, 23 Feb 2021 20:11:23 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233199AbhBWUK6 (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 23 Feb 2021 15:10:58 -0500
Received: from aserp2120.oracle.com ([141.146.126.78]:41266 "EHLO
        aserp2120.oracle.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S234171AbhBWUJK (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 23 Feb 2021 15:09:10 -0500
Received: from pps.filterd (aserp2120.oracle.com [127.0.0.1])
        by aserp2120.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK4IaU091499;
        Tue, 23 Feb 2021 20:08:20 GMT
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=oracle.com;
 h=from : to : cc :
 subject : date : message-id : in-reply-to : references :
 content-transfer-encoding : content-type : mime-version;
 s=corp-2020-01-29; bh=Mo8PBUmkRrVcmhWa92u/HhCvZoMCRaBTR/gdegtOq70=;
 b=WX7EV91cnopkQqxLmA3afZdLfW9evJUD9Xn1F0P2fpfe7RkJSGieek5MusWbiS1HCQtc
 vgxESfPWtxUiKVfwrMu55NGbmMp/ENqLDaKBCXMOBzvGCxG12OaeACILlObzkk3X3Gtg
 6AEhGTzYcty5RJBEAMAIsU7FQ85f/WqBPic58D6AHaJWmyybpzg/KB4Q1JvjAk8CCZxj
 IdFrgvAdNmFmP6/KiIshesEopVPsNTAGBWGyLtilQscBIJ3H/Kq7R5wH7o3/9+haNVlB
 XFSv7vhGzYREWjiMO+lRkRA6Vg3beunc+jVJWy2K08dPyqGZFKPpEvzviMxdYMJ4y7yb Sw==
Received: from aserp3030.oracle.com (aserp3030.oracle.com [141.146.126.71])
        by aserp2120.oracle.com with ESMTP id 36ttcm8pm6-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:20 +0000
Received: from pps.filterd (aserp3030.oracle.com [127.0.0.1])
        by aserp3030.oracle.com (8.16.0.42/8.16.0.42) with SMTP id
 11NK6RGU172790;
        Tue, 23 Feb 2021 20:08:20 GMT
Received: from nam12-bn8-obe.outbound.protection.outlook.com
 (mail-bn8nam12lp2174.outbound.protection.outlook.com [104.47.55.174])
        by aserp3030.oracle.com with ESMTP id 36v9m52nk3-1
        (version=TLSv1.2 cipher=ECDHE-RSA-AES256-GCM-SHA384 bits=256
 verify=OK);
        Tue, 23 Feb 2021 20:08:20 +0000
ARC-Seal: i=1; a=rsa-sha256; s=arcselector9901; d=microsoft.com; cv=none;
 b=d+Oshzw2ussjUyMQs6cYPvdT5yFZh5PZeLnpANEAYflZKv8CP3QsBaz48aAzu9dBFv28sKXFka0VvlUEz4IdTAMQcPm/0y8PLh6YbygsxZoEe3TG+Om1imyWXbP7fkfyvkw3krvaU4dtfgJRJScscK6eeJnK6ghmfI7U2mj/a9hA5KJO7YtaTq7LOg5QUpUWSjkxpSrGyQDVdWHEcPmY501J6N/2u80H97oFdLRvG5jyWyDFGbyzsN9CtgzjEfeMCkn34emF1MIeyd9UxOFOQMqR7foz0kDvEfzErLz5Sz/apr7fLJVfrw0bQvHZMjQqPaqo9SG5v4MNgkQsSXGtgw==
ARC-Message-Signature: i=1; a=rsa-sha256; c=relaxed/relaxed; d=microsoft.com;
 s=arcselector9901;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Mo8PBUmkRrVcmhWa92u/HhCvZoMCRaBTR/gdegtOq70=;
 b=NfTZIBFE7U9SPFhV05mRC62+y6/8EC++h3YmP7jcj/1d0KtDAIYPbHgSlLKBxcpfVWA5HVLUZKocRAX1m/Kz7szrSPW7g1WK8Z/qAJVwAKSTmuQ5r3hWCdp2g3SvMCn1rXtRZn792h6Cr+YHV+oDAorR7Tx2mb9lEFVMzorndyE5dRuHH0WPToLZdhep2vIg+LvB7EmHixed1wSpYKMRXKB63hDfO4Vh23LebgrodnR59eQ72cs8+wY/NgRT/ekCnZrNmARSYiiDijBSinTfFILowddLzkDsbzrEU8mtCDFx51YVshMPW8gEqBxe0eZpy+3mKVLVKdf6jJImnWqJSQ==
ARC-Authentication-Results: i=1; mx.microsoft.com 1; spf=pass
 smtp.mailfrom=oracle.com; dmarc=pass action=none header.from=oracle.com;
 dkim=pass header.d=oracle.com; arc=none
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=oracle.onmicrosoft.com; s=selector2-oracle-onmicrosoft-com;
 h=From:Date:Subject:Message-ID:Content-Type:MIME-Version:X-MS-Exchange-SenderADCheck;
 bh=Mo8PBUmkRrVcmhWa92u/HhCvZoMCRaBTR/gdegtOq70=;
 b=ENH4qhT7e756fg7lQWoVbnzUsN/n9tY7TSX4QFIuS9HRih2xd6Lj4Dud9WvU3+8w0HPfvQNHH5jLLzai+W6cmWPbhJUWNa5OM84xJbFMe5qutY45Dyo/AYu6aNW+Ao/x4Y9bxyGSqjIm5lYw3YqWNti8j3uae5wkJsr69IKOUwA=
Authentication-Results: vger.kernel.org; dkim=none (message not signed)
 header.d=none;vger.kernel.org; dmarc=none action=none header.from=oracle.com;
Received: from SN6PR10MB3021.namprd10.prod.outlook.com (2603:10b6:805:cc::19)
 by SA2PR10MB4556.namprd10.prod.outlook.com (2603:10b6:806:119::8) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3868.30; Tue, 23 Feb
 2021 20:08:18 +0000
Received: from SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7]) by SN6PR10MB3021.namprd10.prod.outlook.com
 ([fe80::1871:3741:cc17:bcf7%7]) with mapi id 15.20.3868.033; Tue, 23 Feb 2021
 20:08:18 +0000
From: Krish Sadhukhan <krish.sadhukhan@oracle.com>
To: kvm@vger.kernel.org
Cc: pbonzini@redhat.com, jmattson@google.com, seanjc@google.com
Subject: [PATCH 4/4 v3] KVM: nSVM: Test effect of host RFLAGS.TF on VMRUN
Date: Tue, 23 Feb 2021 14:19:58 -0500
Message-Id: <20210223191958.24218-5-krish.sadhukhan@oracle.com>
X-Mailer: git-send-email 2.25.4
In-Reply-To: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
References: <20210223191958.24218-1-krish.sadhukhan@oracle.com>
X-Originating-IP: [138.3.200.29]
X-ClientProxiedBy: BY3PR05CA0026.namprd05.prod.outlook.com
 (2603:10b6:a03:254::31) To SN6PR10MB3021.namprd10.prod.outlook.com
 (2603:10b6:805:cc::19)
MIME-Version: 1.0
X-MS-Exchange-MessageSentRepresentingType: 1
Received: from ban25x6uut29.us.oracle.com (138.3.200.29) by
 BY3PR05CA0026.namprd05.prod.outlook.com (2603:10b6:a03:254::31) with
 Microsoft SMTP Server (version=TLS1_2,
 cipher=TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384) id 15.20.3890.10 via Frontend
 Transport; Tue, 23 Feb 2021 20:08:17 +0000
X-MS-PublicTrafficType: Email
X-MS-Office365-Filtering-Correlation-Id: 6101446d-e066-40ff-6639-08d8d836c320
X-MS-TrafficTypeDiagnostic: SA2PR10MB4556:
X-Microsoft-Antispam-PRVS: 
 <SA2PR10MB455660B627309B4277BD2A9781809@SA2PR10MB4556.namprd10.prod.outlook.com>
X-MS-Oob-TLC-OOBClassifiers: OLM:6790;
X-MS-Exchange-SenderADCheck: 1
X-Microsoft-Antispam: BCL:0;
X-Microsoft-Antispam-Message-Info: 
 CYyWUugdRBLQ+Ds+PMUTCrvFdbelgdcvQElyFM7HRD1EHWgOCKAQ16hSFk/uTg5VCRcQK/gCnG+RKOY6R2Qhe3m+TIucaLr6f41DA4OpqT6XzGYS28LezF31pMeZCDi81XNfOJnR5rTfDSU+3uAiC2NgLvMngO6kojRn9GF8CpGIZfmVSGqTkWp9H9e80Pr34HxorJCw6OhfWl1NORSCDSO1cXZ/bVWWnN1JTYGTVCS/yd4ml1c3WVnh4E4CbRWEEo0T6xBvIXhRihdRZAngBVdNJJJ1mU457/F8pkUYz4CKXNLIs+rDyj3cHYoHjZu2wAgVa2EAoHMGVn9PJtjd4zgGEGQjH1EPLujEhG4bhg0lvsONiCRhklVQeDrAfVOgrCb4a93x1QNxu5ierXIAzQOinbLvIKCZqq8QcCt5v3tgj7jmsnfbQhcdFLDpZQEVBcDvTRH8x+tuVZq+Q7+BvV/iZ2a/tmUaALcSs0RetpsOurFk+qER4g680Z9quhLJHSYyQ747YYUn0A3dkj5hMQ==
X-Forefront-Antispam-Report: 
 CIP:255.255.255.255;CTRY:;LANG:en;SCL:1;SRV:;IPV:NLI;SFV:NSPM;H:SN6PR10MB3021.namprd10.prod.outlook.com;PTR:;CAT:NONE;SFS:(366004)(376002)(396003)(39860400002)(346002)(136003)(4326008)(6666004)(6916009)(6486002)(16526019)(2906002)(5660300002)(478600001)(66946007)(66476007)(66556008)(186003)(26005)(8936002)(2616005)(36756003)(956004)(83380400001)(1076003)(8676002)(86362001)(44832011)(316002)(52116002)(7696005);DIR:OUT;SFP:1101;
X-MS-Exchange-AntiSpam-MessageData: 
 A7fdJkC4q8DwCl5413mY6zNIbg9l7OeiUZUWDKXOEq/iEJTT6BvbWTy8edHYmAXZzbm8cuual4iRsxwx9KcJsDioS9qmhghxgsLTwWP+3vVeMKyghP80mfuJwjJNasSWFi9iwYA9r4rgirumA9KsMRD56d4QKKp7OTpv2SJJPFKhWEzTAo5P5eKP6tyMhrFofJkiyVzr2WXewsi5QL4OALSnmTj2VpNVaQesm9BAsSnBGOXJY/sMJWSPvVHjlGN2TPixy88z94CTLdTjMuD41W89LISyRsgCLWdkADq025Ud800HWCK762Ttu3plegHkOyb5xQWVW15SREbO3Rj3IGsKFL+0KRZRnlQYwDJlTaqHm0e6sh7ng/tk96GnzUuS50PS+Fyfd+7fWZ0c/mQ27cWhdF+tFvJKhJM64FJAEbqQb5wwQPDp0MO64K+8ttP88qqyWAxX7/yHJoXh+9ms1ImfEccCYhSkRl7TO/xU26mLlzbgkPCcFtA4hf6GUjkrPXJ9lPc6Ua+TXWLZcGtVgUjS0QWpquw0qik0IwXqMV7tk95NqebvZX2lQjtjtdlkswxNV5hhA/x8149uMvRY6CEZTFpUu8MBS3XkhDBW9mjZjbNlfbRXdAgA8khCCXlSX8hiHrDmKxAvdxrz2z24dllQdz5JxKWt7VLj3VEkOaGH19uSj0NQQzyoaQB2/gzX0otUjjxjv5hw444HQzyuw41BCqhFUWqd1ShmPqqZYmdPh1KRCZiAN5sUbENpEDGHH5O3O1QAWvRU2fEgJPqvFBqj0DogHmsReToKXMfkuX71mHEOXSEgokxWD38wyC3i07+fM2HzTyPu9V70ZGlluKMTFldpTNl0XjNOQb0xCb203RVVDnzYNrBcQ/6hbGLIVUMaYa1mmUD3ewwwM+GOw8WnA7dTfGn/lLUNjFi+/afZNGxLVhN10QD40eCrP1qSMJUFUTxDPkToQHVJVTNWpWGVX/KquxqLfbSNEZ4l1BtmDfh4H/O2VNmumdk7xgjt9XC5dcVH7gkDf549RnWeSxYkXQU8NBXfi81aZeR/v8/VyezUAwv7TN/opRMD/KmXhmcDHNXd8wCkD33cp+LIsovI64jTtW9eYcFHAE+N/td1YwX1IwXidH2NXnrZaJDV2Uogxye1FuSqy2XomVh5ORjMSgXtlKdQfybhphC5fni2nEmhYI3su8UHdAfpcfo4soPsdarh+9+gcIbNJZvIvBIfiyTSttfDNKW+AKtX4XiBrib2gnUjfLEwpxeyV0qFXyJvVki9Jj1/6ttwvJ4feyHa8K8hOt21oxQXQjwk1J3FE64YfKmqkt+g0wtnXrW4
X-OriginatorOrg: oracle.com
X-MS-Exchange-CrossTenant-Network-Message-Id: 
 6101446d-e066-40ff-6639-08d8d836c320
X-MS-Exchange-CrossTenant-AuthSource: SN6PR10MB3021.namprd10.prod.outlook.com
X-MS-Exchange-CrossTenant-AuthAs: Internal
X-MS-Exchange-CrossTenant-OriginalArrivalTime: 23 Feb 2021 20:08:18.8837
 (UTC)
X-MS-Exchange-CrossTenant-FromEntityHeader: Hosted
X-MS-Exchange-CrossTenant-Id: 4e2c6054-71cb-48f1-bd6c-3a9705aca71b
X-MS-Exchange-CrossTenant-MailboxType: HOSTED
X-MS-Exchange-CrossTenant-UserPrincipalName: 
 sbinGBnX6YkfHt4JucPZo2WnSYOdV7FD+let18nM2rkM3av6HhLlHTl+IeIG8ofCvPWrcAtLVljrDxwlLXUZFlE4IdbSCQPgr2flPUAB7oE=
X-MS-Exchange-Transport-CrossTenantHeadersStamped: SA2PR10MB4556
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 malwarescore=0
 adultscore=0
 suspectscore=0 mlxlogscore=999 mlxscore=0 spamscore=0 bulkscore=0
 phishscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
X-Proofpoint-Virus-Version: vendor=nai engine=6200 definitions=9904
 signatures=668683
X-Proofpoint-Spam-Details: rule=notspam policy=default score=0 mlxlogscore=999
 adultscore=0
 lowpriorityscore=0 spamscore=0 mlxscore=0 bulkscore=0 clxscore=1015
 priorityscore=1501 malwarescore=0 impostorscore=0 suspectscore=0
 phishscore=0 classifier=spam adjust=0 reason=mlx scancount=1
 engine=8.12.0-2009150000 definitions=main-2102230169
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

According to section "VMRUN and TF/RF Bits in EFLAGS" in AMD APM vol 2,

    "From the host point of view, VMRUN acts like a single instruction,
     even though an arbitrary number of guest instructions may execute
     before a #VMEXIT effectively completes the VMRUN. As a single
     host instruction, VMRUN interacts with EFLAGS.TF like ordinary
     instructions. EFLAGS.TF causes a #DB trap after the VMRUN completes
     on the host side (i.e., after the #VMEXIT from the guest).

Suggested-by: Paolo Bonzini <pbonzini@redhat.com>
Signed-off-by: Krish Sadhukhan <krish.sadhukhan@oracle.com>
---
 x86/svm_tests.c | 115 ++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 115 insertions(+)

diff --git a/x86/svm_tests.c b/x86/svm_tests.c
index 29a0b59..466a13c 100644
--- a/x86/svm_tests.c
+++ b/x86/svm_tests.c
@@ -2002,6 +2002,118 @@ static bool init_intercept_check(struct svm_test *test)
     return init_intercept;
 }
 
+/*
+ * Setting host EFLAGS.TF causes a #DB trap after the VMRUN completes on the
+ * host side (i.e., after the #VMEXIT from the guest).
+ *
+ * [AMD APM]
+ */
+static volatile u8 host_rflags_guest_main_flag = 0;
+static volatile u8 host_rflags_db_handler_flag = 0;
+static volatile bool host_rflags_ss_on_vmrun = false;
+static volatile bool host_rflags_vmrun_reached = false;
+static volatile bool host_rflags_set_tf = false;
+static u64 post_vmrun_rip;
+
+extern void *vmrun_rip;
+
+static void host_rflags_db_handler(struct ex_regs *r)
+{
+	if (host_rflags_ss_on_vmrun) {
+		if (host_rflags_vmrun_reached) {
+			r->rflags &= ~X86_EFLAGS_TF;
+			post_vmrun_rip = r->rip;
+		} else {
+			if (r->rip == (u64)(&vmrun_rip))
+				host_rflags_vmrun_reached = true;
+		}
+	} else {
+		r->rflags &= ~X86_EFLAGS_TF;
+	}
+}
+
+static void host_rflags_prepare(struct svm_test *test)
+{
+	default_prepare(test);
+	handle_exception(DB_VECTOR, host_rflags_db_handler);
+	set_test_stage(test, 0);
+}
+
+static void host_rflags_prepare_gif_clear(struct svm_test *test)
+{
+	if (host_rflags_set_tf)
+		write_rflags(read_rflags() | X86_EFLAGS_TF);
+}
+
+static void host_rflags_test(struct svm_test *test)
+{
+	while (1) {
+		if (get_test_stage(test) > 0 && host_rflags_set_tf &&
+		    (!host_rflags_ss_on_vmrun) &&
+		    (!host_rflags_db_handler_flag))
+			host_rflags_guest_main_flag = 1;
+		vmmcall();
+		if (get_test_stage(test) == 3)
+			break;
+	}
+}
+
+static bool host_rflags_finished(struct svm_test *test)
+{
+	switch (get_test_stage(test)) {
+	case 0:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL) {
+			report(false, "Unexpected VMEXIT. Exit reason 0x%x",
+			    vmcb->control.exit_code);
+			return true;
+		}
+		vmcb->save.rip += 3;
+		/*
+		 * Setting host EFLAGS.TF not immediately before VMRUN, causes
+		 * #DB trap before first guest instruction is executed
+		 */
+		host_rflags_set_tf = true;
+		break;
+	case 1:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL ||
+		    (!host_rflags_guest_main_flag)) {
+			report(false, "Unexpected VMEXIT or #DB handler"
+			    " invoked before guest main. Exit reason 0x%x",
+			    vmcb->control.exit_code);
+			return true;
+		}
+		vmcb->save.rip += 3;
+		/*
+		 * Setting host EFLAGS.TF immediately before VMRUN, causes #DB
+		 * trap after VMRUN completes on the host side (i.e., after
+		 * VMEXIT from guest).
+		 */
+		host_rflags_ss_on_vmrun = true;
+		break;
+	case 2:
+		if (vmcb->control.exit_code != SVM_EXIT_VMMCALL ||
+		    post_vmrun_rip - (u64)(&vmrun_rip) != 3) {
+			report(false, "Unexpected VMEXIT or RIP mismatch."
+			    " Exit reason 0x%x, VMRUN RIP: %lx, post-VMRUN"
+			    " RIP: %lx", vmcb->control.exit_code,
+			    (u64)(&vmrun_rip), post_vmrun_rip);
+			return true;
+		}
+		host_rflags_set_tf = false;
+		vmcb->save.rip += 3;
+		break;
+	default:
+		return true;
+	}
+	inc_test_stage(test);
+	return get_test_stage(test) == 4;
+}
+
+static bool host_rflags_check(struct svm_test *test)
+{
+	return get_test_stage(test) == 3;
+}
+
 #define TEST(name) { #name, .v2 = name }
 
 /*
@@ -2492,6 +2604,9 @@ struct svm_test svm_tests[] = {
     { "svm_init_intercept_test", smp_supported, init_intercept_prepare,
       default_prepare_gif_clear, init_intercept_test,
       init_intercept_finished, init_intercept_check, .on_vcpu = 2 },
+    { "host_rflags", default_supported, host_rflags_prepare,
+      host_rflags_prepare_gif_clear, host_rflags_test,
+      host_rflags_finished, host_rflags_check },
     TEST(svm_cr4_osxsave_test),
     TEST(svm_guest_state_test),
     TEST(svm_vmrun_errata_test),
